### DISCUSSION TITLE
Keep values given to a listen-only device?

### DISCUSSION THREAD

**Discussion by LongnoseRob** [pymeasure/pymeasure#721] [Q&A] [answered]
Hello all,

I am working at the moment on an implementation of a [device (HP8657B signal generator)](https://github.com/LongnoseRob/pymeasure/tree/HP8657B).

This device ihas some vintage, but can be programmed via HPIB/GPIB, hpwever it is a listen-only device, thus I can set the e.g. the frequency but cannot read anything back formt he device as the firmware does not offer any quiery commands.

I was wondering if there would be an **easy** possibility to have properties of `Instrument.setting()` "keep" the values asssgined, so that I could read the value later back later during the script.

code example:
```
from pymeasure.instruments.hp import HP8657B
siggen.output_enabled = False
siggen.frequency = 1.2345e9
siggen.amplitude = -30
..

siggen.output_enabled = True
..
print(siggen.frequency)  # this faults normally with LookupError: Instrument property can not be read.
```
If this is not possible with an easy hack, than I am fine with it, just corious..

**Comment by BenediktBurger** [comment]
Normally, we don't want to store the settings set by the computer, as a user input at the device itself (e.g. changing the limit from volts to amps), might change the settings creating a difference between the stored (still volts) and the actual setting (amps). Therefore, it is always better to read the current configuration.

If you use normal python properties, you can store these values. You should think twice, though, whether it justifies the possible complication and problems.

**Comment by mcdo0486** [comment, marked as answer]
Hey,

I ran into that when implementing the HP 4192A. There isn't a way to query for the current settings of the instrument, but [on power on the settings are in a known state. ](https://github.com/mcdo0486/pymeasure/blob/93f408fc24817ce29194cdd2f193f163ed1bdea1/pymeasure/instruments/hp/hp4192A.py#L30)

In the HP4192A instrument class I [overrode the setting static method](https://github.com/mcdo0486/pymeasure/blob/93f408fc24817ce29194cdd2f193f163ed1bdea1/pymeasure/instruments/hp/hp4192A.py#L82) of the Instrument parent class so that `fget` would return something from `self.settings` and add `self_validator` argument to validate stuff based on the current `self.settings` state.

I'm not sure if that fulfills your definition of *easy* though. 

The branch and instrument driver linked is still under development.
https://github.com/mcdo0486/pymeasure/blob/hp4192A/pymeasure/instruments/hp/hp4192A.py

**Reply by LongnoseRob** [reply]
Thanks for the reply and the reference to the 4192A
